src/main/kotlin/org/rust/lang/doc/psi/RsDocLinkDestinationParser.kt

```diff
@@ -0,0 +1,171 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.doc.psi
+
+import com.intellij.lang.LanguageParserDefinitions
+import com.intellij.lang.impl.PsiBuilderImpl
+import com.intellij.lang.parser.GeneratedParserUtilBase
+import com.intellij.lang.parser.GeneratedParserUtilBase._NONE_
+import com.intellij.psi.TokenType
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import com.intellij.psi.impl.source.tree.TreeElement
+import com.intellij.util.CharTable
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.lexer.RsLexer
+import org.rust.lang.core.parser.RustParser
+import org.rust.lang.doc.psi.RsDocElementTypes.DOC_LINK_DEFINITION
+
+/**
+ * `fn@mod1::mod2::func#anchor`
+ *  ~~~ prefix         ~~~~~~~ suffix
+ *     ~~~~~~~~~~~~~~~~ text
+ */
+private class LinkInfo private constructor(
+    val text: CharSequence,
+    val prefixLength: Int,
+    val suffixLength: Int,
+) {
+    constructor(text: CharSequence) : this(text, prefixLength = 0, suffixLength = 0)
+
+    fun subSequence(start: Int, end: Int): LinkInfo {
+        if (start == 0 && end == text.length) return this
+        check(start in 0..end && end <= text.length)
+        val newText = text.subSequence(start, end)
+        return LinkInfo(newText, prefixLength + start, suffixLength + (text.length - end))
+    }
+
+    fun removePrefix(length: Int): LinkInfo = subSequence(length, text.length)
+    fun removeSuffix(length: Int): LinkInfo = subSequence(0, text.length - length)
+}
+
+/**
+ * See `preprocess_link` in `src/librustdoc/passes/collect_intra_doc_links.rs`
+ * Algorithm:
+ * - Remove backticks
+ * - Remove hash suffix
+ * - Remove disambiguator prefix/suffix
+ * - Parse path using usual parser (it will parse generics if any)
+ */
+object RsDocLinkDestinationParser {
+    fun parse(text: CharSequence, charTable: CharTable): TreeElement {
+        val info = createLinkInfo(text) ?: return docDataLeaf(text, charTable)
+        return parsePathAndCreateNodes(text, charTable, info) ?: docDataLeaf(text, charTable)
+    }
+
+    private fun createLinkInfo(text: CharSequence): LinkInfo? {
+        if (text.isEmpty() || text.contains("/")) return null
+
+        return LinkInfo(text)
+            .trimBackticks()
+            ?.removeHashAnchor()
+            ?.removeDisambiguator()
+            ?.takeIf { canBeCorrectLink(it.text) }
+    }
+
+    // "`func`" -> "func"
+    // Remove any number of backticks at the beginning and end.
+    // Rustdoc also removes backticks in the middle, but we don't support it
+    private fun LinkInfo.trimBackticks(): LinkInfo? {
```

Comment:
I'd prefer moving these extension on top-level near `LinkInfo` or into `LinkInfo` itself to make more explicit that they don't use any context outside of `LinkInfo`

Reply:
Do you mean only `trimBackticks` or also `removeHashAnchor` etc? Actually `RsDocLinkDestinationParser` is `object`, so there can't be any context anyway

Reply:
Nevermind. I thought a little better and now it seems to me that the way it is even better.